sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala

```diff
@@ -0,0 +1,383 @@
+      createFieldWriter(vector)
+    }
+    new ArrowWriter(root, children.toArray)
+  }
+
+  private def createFieldWriter(vector: ValueVector): ArrowFieldWriter = {
+    val field = vector.getField()
+    ArrowUtils.fromArrowField(field) match {
+      case BooleanType =>
+        new BooleanWriter(vector.asInstanceOf[NullableBitVector])
+      case ByteType =>
+        new ByteWriter(vector.asInstanceOf[NullableTinyIntVector])
+      case ShortType =>
+        new ShortWriter(vector.asInstanceOf[NullableSmallIntVector])
+      case IntegerType =>
+        new IntegerWriter(vector.asInstanceOf[NullableIntVector])
+      case LongType =>
+        new LongWriter(vector.asInstanceOf[NullableBigIntVector])
+      case FloatType =>
+        new FloatWriter(vector.asInstanceOf[NullableFloat4Vector])
+      case DoubleType =>
+        new DoubleWriter(vector.asInstanceOf[NullableFloat8Vector])
+      case StringType =>
+        new StringWriter(vector.asInstanceOf[NullableVarCharVector])
+      case BinaryType =>
+        new BinaryWriter(vector.asInstanceOf[NullableVarBinaryVector])
+      case ArrayType(_, _) =>
+        val v = vector.asInstanceOf[ListVector]
+        val elementVector = createFieldWriter(v.getDataVector())
+        new ArrayWriter(v, elementVector)
+      case StructType(_) =>
+        val v = vector.asInstanceOf[NullableMapVector]
+        val children = (0 until v.size()).map { ordinal =>
+          createFieldWriter(v.getChildByOrdinal(ordinal))
+        }
+        new StructWriter(v, children.toArray)
+      case dt =>
+        throw new UnsupportedOperationException(s"Unsupported data type: ${dt.simpleString}")
+    }
+  }
+}
+
+class ArrowWriter(
+    val root: VectorSchemaRoot,
+    fields: Array[ArrowFieldWriter]) {
+
+  def schema: StructType = StructType(fields.map { f =>
+    StructField(f.name, f.dataType, f.nullable)
+  })
+
+  private var count: Int = 0
+
+  def write(row: InternalRow): Unit = {
+    var i = 0
+    while (i < fields.size) {
+      fields(i).write(row, i)
+      i += 1
+    }
+    count += 1
+  }
+
+  def finish(): Unit = {
+    root.setRowCount(count)
+    fields.foreach(_.finish())
+  }
+
+  def reset(): Unit = {
+    root.setRowCount(0)
+    count = 0
+    fields.foreach(_.reset())
+  }
+}
+
+private[arrow] abstract class ArrowFieldWriter {
+
+  def valueVector: ValueVector
+  def valueMutator: ValueVector.Mutator
+
+  def name: String = valueVector.getField().getName()
+  def dataType: DataType = ArrowUtils.fromArrowField(valueVector.getField())
+  def nullable: Boolean = valueVector.getField().isNullable()
+
+  def setNull(): Unit
+  def setValue(input: SpecializedGetters, ordinal: Int): Unit
+  def skip(): Unit
+
+  protected var count: Int = 0
+
+  def write(input: SpecializedGetters, ordinal: Int): Unit = {
+    if (input.isNullAt(ordinal)) {
+      setNull()
+    } else {
+      setValue(input, ordinal)
+    }
+    count += 1
+  }
+
+  def writeSkip(): Unit = {
+    skip()
+    count += 1
```

Comment:
For skipping purpose, is it enough to just do `count += 1`? e.g. `vector.set(1, v1); vector.set(3, v3)`, value 2 is skipped.